### PR TITLE
feat(serve): default MCP review to diff scope (changed-set vs main)

### PR DESCRIPTION
**Pull request by kodflow**
## Summary

Make the ktn-linter MCP daemon default to **diff-scoped review**: it surfaces
only issues on code changed vs the default branch, with an automatic loud
full-scan fallback when no git base resolves (no `.git`, shallow clone,
unresolved default branch). Configured globally once via `review_scope: diff | full`
(default `diff`) — never per-rule.

## What changed

- **`pkg/git`** (new): changed-set resolver — three-dot merge-base + staged +
  unstaged + untracked, rename/delete/generated/shallow handling, per-file line
  ranges. Shells out to git (no go-git dep), reuses `pkg/cache.IsGeneratedFile`.
- **`pkg/prompt/diff_granularity.go`**: `ClassifyDiffGranularity` (line / file /
  package). Default `file` (safe-coarse, never silently drops), `package` for
  aggregate/cross-file rules (PKG-SIZE, COVERAGE, TEST-FILES, unused-symbol),
  `line` for statement-local rules.
- **`pkg/serve/diff_scope.go`**: `applyDiffScope` — response-time tag (`in_diff`)
  + hide; the analysis cache always holds the full set (pure decoration,
  parallel to `in_active_set`). Wired into the MCP `scan` tool, HTTP `/scan`,
  and the post-tool-use hook. Envelope gains `review_scope` / `diff_base` /
  `diff_head` / `diff_warning`.
- **`pkg/config`**: global `review_scope` field (mirrors the `surface_strategy`
  Raw + resolver + validator pattern).
- **phases/severity**: moved `KTN-TEST-SUFFIX/PKGNAME/FILES/PLACEMENT` phase 1 →
  8 (WARNING) so a diff review of production code is not gated on test-file
  discipline.
- docs: README, `.ktn-linter.yaml`, 4× CLAUDE.md.

## Design invariants

- Filtering happens at **response time** only — the content-addressed cache is
  never poisoned with a diff-filtered subset.
- **Fail-safe, loud**: unresolved base / shallow / no `.git` → full scan +
  `diff_warning`, never a silent empty diff.
- Package-scoped predicate uses touched-package-dirs (not file membership) so a
  missing-test diagnostic on a new file still surfaces — the #1 false-negative
  guard, covered by an integration test.

## Verification

- `make build` ✓ · `make lint` ✅ 0 issues project-wide · `make validate` ✓
- 4959 tests pass across affected packages (the only `make test` red is a
  pre-existing unix-socket path-length limitation, environmental).

- Diff-scoped default with safe full-scan fallback: introduced global review_scope (diff|full, default diff). New pkg/git.Resolve computes ChangedSet (merge-base + staged/unstaged/untracked, handles renames/deletes, skips generated files) and returns ResolutionValue; any git/base resolution or collection error degrades loudly to FullFallback with a diff_warning. Filtering is applied at response time only; the analysis/cache continue to hold full issue sets.

- Per-rule granularity, phase & severity updates: added pkg/prompt.ClassifyDiffGranularity (GranularityFile/Line/Package) and rule lookup maps to control when diagnostics are shown under diff scope. Reclassified KTN-TEST-* rules into phase 8 (tests) and changed KTN-TEST-SUFFIX severity from ERROR → WARNING to avoid gating diff reviews; no other KTN rule codes were added or removed.

- Serve/MCP/CLI and wire-format changes: pkg/serve stamps IssueMessage.InDiff and filters surfaced issues by resolved scope; ScanRequest/ScanResponse/ScanToolOutputMessage extended with review_scope, diff_base, diff_head, diff_warning. MCP proxy path now forwards per-request scope and returned diff metadata. CLI lint adds a daemon-guard probe (with --force / KTN_FORCE_LOCAL override), prints a loud override message when forced, and introduces a new DaemonActive exit code to avoid clobbering a running MCP daemon.

- Tests and validation: substantial new coverage across pkg/git (parsing, exec, resolve with real temp repos), pkg/prompt (granularity and phase tests), and pkg/serve (diff-scope unit + git-backed integration tests). Config validation added (RawReviewScope + ReviewScope() accessor and ErrInvalidReviewScope) and Resolve now propagates git exec failures so fallback behavior is tested. Affected packages build/lint cleanly (≈4,959 tests pass across affected packages); one unrelated environment-specific socket-bind test failure remains.

- Govet/modernize wrappers: no govet/modernize-style wrapper or automated code-modernization changes were introduced in this PR.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -92,6 +97,19 @@ type Config struct {
...
+	//
+	// Recognized values:
+	//   - "diff" (default): surface only issues on code changed vs the main
+	//     branch; automatically falls back to "full" when there is no
+	//     resolvable git base (no .git, shallow clone, missing default branch).
+	//   - "full":           surface every issue across the whole project.
+	//
+	// Global toggle only — there is intentionally no per-rule override.
+	// Use the ReviewScope() method to get the resolved value.
+	RawReviewScope string `yaml:"review_scope,omitempty"`
```
[the remark this answers is not in the document]
Comment (kodflow) [reply pkg/config/config.go:111]: Addressed in 97bacdf9: `Config.Merge()` now propagates `RawReviewScope` (empty = not-set-in-this-layer, higher-precedence non-empty wins). The AI-prompt's secondary suggestion to also touch `SerializationTags()` was not applied — that method returns the KTN-STRUCT-JSONTAG detection tag set (json/yaml/xml…), not config field tags, so it is unrelated to review_scope round-trip.

File: pkg/git/diff_parse.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+			}
+			continue
+		}
+		//: Otherwise the line may update the pre/post-edit path bookkeeping.
+		blk.target, blk.oldPath = updatePaths(line, blk.target, blk.oldPath)
+	}
+
+	//: Flush the final file block.
+	flushFile(set, root, gen, blk)
+}
```
[the remark this answers is not in the document]
Comment (kodflow) [reply pkg/git/diff_parse.go:70]: Addressed in 97bacdf9: path-header parsing is now gated to the pre-hunk region via an `inHunks` flag, so a -U0 hunk-body line such as `-- a/x.go` (rendered `--- a/x.go`) or `++ b/x.go` is treated as content, never a path header. Added a regression test (hunk body lines are not misparsed as path headers).

File: pkg/serve/scan_tool_output.go [hunk 3]
```diff
@@ -14,4 +14,13 @@ type ScanToolOutputMessage struct {
...
 	Err string `json:"err,omitempty"`
+	//: ReviewScope is "diff" or "full" — the resolved review scope. In diff
+	//: mode only issues on code changed vs the default branch are returned;
+	//: DiffBase/DiffHead identify the comparison commits and DiffWarning is
+	//: non-empty when diff mode degraded to a full scan (no .git, shallow
+	//: clone, unresolved base). All omitempty for wire compatibility.
+	ReviewScope string `json:"review_scope,omitempty"`
+	DiffBase    string `json:"diff_base,omitempty"`
+	DiffHead    string `json:"diff_head,omitempty"`
+	DiffWarning string `json:"diff_warning,omitempty"`
```
[the remark this answers is not in the document]
Comment (kodflow) [reply pkg/serve/scan_tool_output.go:25]: Addressed in 97bacdf9: `runProxyScan` now forwards `ReviewScope`, `DiffBase`, `DiffHead`, `DiffWarning` (and `Err`) from the proxied `ScanResponse`, so proxy mode is schema-compatible with the direct `executeScanTool` path. The remote daemon already applies the scope filter to `resp.Issues`, so the proxy only forwards the metadata.

File: pkg/severity/severity.go [hunk 4]
```diff
@@ -252,7 +252,7 @@ var (
 		{code: "KTN-STRUCT-SPLITPREFIX", level: SeverityInfo},         // File exceeds size thresholds — split into siblings
 
 		// TEST - Tests (47 entries)
```
[the remark this answers is not in the document]
Comment (kodflow) [reply pkg/severity/severity.go]: Addressed in 97bacdf9: TEST section comment corrected 47 -> 48 to match the registered KTN-TEST-* rules.
